pytorch_lightning/trainer/training_loop.py

```diff
@@ -685,69 +702,79 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
 
         for split_idx, split_batch in enumerate(splits):
 
-            # create an iterable for optimizers and loop over them
-            for opt_idx, optimizer in optimizers:
-
-                # toggle model params + set info to logger_connector
-                self.run_train_split_start(split_idx, split_batch, opt_idx, optimizer)
-
-                result = AttributeDict()
-                if self.should_accumulate():
-                    # For gradient accumulation
+            if self.trainer.lightning_module.automatic_optimization:
+                for opt_idx, optimizer in optimizers:
+                    if self._should_skip_optimizer(opt_idx, batch_idx):
+                        # frequency of this optimizer does not align with current batch index, skip it
+                        continue
+                    result = self.run_batch_split(batch_idx, split_idx, split_batch, opt_idx, optimizer)
+                    if result:
+                        batch_outputs[opt_idx].append(result.training_step_output_for_epoch_end)
+                        grad_norm_dict = result.get("grad_norm_dict", {})
+            else:
+                result = self.run_batch_split(batch_idx, split_idx, split_batch)
+                if result:
+                    # this if check is required for
+                    # tests/trainer/optimization/test_manual_optimization.py::test_step_with_optimizer_closure_and_accumulated_grad
+                    # TODO: make grad accumulation + manual optimization incompatible to simplify this logic here!
+                    batch_outputs[0].append(result.training_step_output_for_epoch_end)
```

Comment:
Here I'm separating 
1. automatic optimization which has to loop over the optimizers
2. manual optimization which should NOT loop over optimizers

previously the two were mashed together under one optimizer loop where in the special case of manual optimizatio, the loop contained a single optimizer for some reason. 

By splitting it, we are now able to better understand what the return value for the results look like and simplifies further refactors!